Multiple issues with SmallRye GraphQL Client

The quickstart for SmallRye GraphQL Client doesn't work anymore.

There are multiple issues.

## Config

The config that was used looks like it's not supported anymore as I get:

```
Caused by: java.util.NoSuchElementException: SRGQLDC035001: URL not configured for client named 'star-wars-dynamic'. Please define the property smallrye.graphql.client.CLIENT_NAME/url
```

I tried to follow this advice but it didn't work so the advice is wrong or something is broken.

I was able to get it to work by using the Quarkus way (following what @jmartisk did in his config overhaul PR). See the PR here: https://github.com/quarkusio/quarkus-quickstarts/pull/897 .

In a perfect world, that would be the syntax recommended by the error message (but given the error is triggered by SmallRye GraphQL Client, it might be a bit tricky).

## Native not working

I then tried to compile to native and got:

```
2021-07-06 17:20:27,924 ERROR [io.qua.run.Application] (main) Failed to start application (with profile prod): java.lang.NullPointerException
	at io.smallrye.graphql.client.GraphQLClientsConfiguration.initialize(GraphQLClientsConfiguration.java:49)
	at io.smallrye.graphql.client.GraphQLClientsConfiguration_Bean.create(GraphQLClientsConfiguration_Bean.zig:144)
	at io.smallrye.graphql.client.GraphQLClientsConfiguration_Bean.create(GraphQLClientsConfiguration_Bean.zig:160)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:96)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:29)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:26)
```

Do we really test native properly in the main repo?

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -46,7 +46,7 @@
         <smallrye-health.version>3.1.1</smallrye-health.version>
         <smallrye-metrics.version>3.0.1</smallrye-metrics.version>
         <smallrye-open-api.version>2.1.7</smallrye-open-api.version>
-        <smallrye-graphql.version>1.2.7</smallrye-graphql.version>
+        <smallrye-graphql.version>1.2.8</smallrye-graphql.version>
         <smallrye-opentracing.version>2.0.0</smallrye-opentracing.version>
         <smallrye-fault-tolerance.version>5.1.0</smallrye-fault-tolerance.version>
         <smallrye-jwt.version>3.2.0</smallrye-jwt.version>

--- a/extensions/smallrye-graphql-client/deployment/src/test/java/io/quarkus/smallrye/graphql/client/deployment/DynamicGraphQLClientInjectionTest.java
+++ b/extensions/smallrye-graphql-client/deployment/src/test/java/io/quarkus/smallrye/graphql/client/deployment/DynamicGraphQLClientInjectionTest.java
@@ -35,13 +35,13 @@ public class DynamicGraphQLClientInjectionTest {
     static QuarkusUnitTest test = new QuarkusUnitTest()
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
                     .addClasses(TestingGraphQLApi.class, Person.class)
-                    .addAsResource(new StringAsset("people/mp-graphql/url=" + url + "\n" +
-                            "people/mp-graphql/header/My-Header=My-Value"),
+                    .addAsResource(new StringAsset("people-client/mp-graphql/url=" + url + "\n" +
+                            "people-client/mp-graphql/header/My-Header=My-Value"),
                             "application.properties")
                     .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml"));
 
     @Inject
-    @GraphQLClient("people")
+    @GraphQLClient("people-client")
     DynamicGraphQLClient client;
 
     @Test

--- a/extensions/smallrye-graphql-client/runtime/src/main/java/io/quarkus/smallrye/graphql/client/runtime/GraphQLClientConfig.java
+++ b/extensions/smallrye-graphql-client/runtime/src/main/java/io/quarkus/smallrye/graphql/client/runtime/GraphQLClientConfig.java
@@ -1,6 +1,7 @@
 package io.quarkus.smallrye.graphql.client.runtime;
 
 import java.util.Map;
+import java.util.Optional;
 
 import io.quarkus.runtime.annotations.ConfigGroup;
 import io.quarkus.runtime.annotations.ConfigItem;
@@ -12,7 +13,7 @@ public class GraphQLClientConfig {
      * The URL location of the target GraphQL service.
      */
     @ConfigItem
-    public String url;
+    public Optional<String> url;
 
     /**
      * HTTP headers to add when communicating with the target GraphQL service.

--- a/extensions/smallrye-graphql-client/runtime/src/main/java/io/quarkus/smallrye/graphql/client/runtime/GraphQLClientConfigurationMergerBean.java
+++ b/extensions/smallrye-graphql-client/runtime/src/main/java/io/quarkus/smallrye/graphql/client/runtime/GraphQLClientConfigurationMergerBean.java
@@ -46,14 +46,12 @@ public class GraphQLClientConfigurationMergerBean {
             if (!upstreamConfiguration.getClients().containsKey(configKey)) {
                 GraphQLClientConfiguration transformed = new GraphQLClientConfiguration();
                 transformed.setHeaders(quarkusConfig.headers);
-                transformed.setUrl(quarkusConfig.url);
+                quarkusConfig.url.ifPresent(transformed::setUrl);
                 upstreamConfiguration.getClients().put(configKey, transformed);
             } else {
                 // if SmallRye configuration already contains this client, override it with the Quarkus configuration
                 GraphQLClientConfiguration upstreamConfig = upstreamConfiguration.getClients().get(configKey);
-                if (quarkusConfig.url != null) {
-                    upstreamConfig.setUrl(quarkusConfig.url);
-                }
+                quarkusConfig.url.ifPresent(upstreamConfig::setUrl);
                 // merge the headers
                 if (quarkusConfig.headers != null) {
                     upstreamConfig.getHeaders().putAll(quarkusConfig.headers);

--- a/extensions/smallrye-graphql-client/runtime/src/main/java/io/quarkus/smallrye/graphql/client/runtime/SmallRyeGraphQLClientRecorder.java
+++ b/extensions/smallrye-graphql-client/runtime/src/main/java/io/quarkus/smallrye/graphql/client/runtime/SmallRyeGraphQLClientRecorder.java
@@ -30,7 +30,7 @@ public class SmallRyeGraphQLClientRecorder {
                 throw new RuntimeException(e);
             }
         }).collect(Collectors.toList());
-        configBean.apiClasses(classes);
+        configBean.apiClasses(classes, true);
     }
 
     public RuntimeValue<GraphQLClientSupport> clientSupport(Map<String, String> shortNamesToQualifiedNames) {

--- a/integration-tests/smallrye-graphql-client/pom.xml
+++ b/integration-tests/smallrye-graphql-client/pom.xml
@@ -21,11 +21,14 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-smallrye-graphql-client</artifactId>
         </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy-deployment</artifactId>
+        </dependency>
 
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-rest-client-deployment</artifactId>
-            <scope>test</scope>
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>

--- a/integration-tests/smallrye-graphql-client/src/main/java/io/quarkus/io/smallrye/graphql/client/GraphQLClientTester.java
+++ b/integration-tests/smallrye-graphql-client/src/main/java/io/quarkus/io/smallrye/graphql/client/GraphQLClientTester.java
@@ -1,0 +1,119 @@
+package io.quarkus.io.smallrye.graphql.client;
+
+import static io.smallrye.graphql.client.core.Argument.arg;
+import static io.smallrye.graphql.client.core.Argument.args;
+import static io.smallrye.graphql.client.core.Document.document;
+import static io.smallrye.graphql.client.core.Field.field;
+import static io.smallrye.graphql.client.core.Operation.operation;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+
+import io.smallrye.graphql.client.Response;
+import io.smallrye.graphql.client.core.Document;
+import io.smallrye.graphql.client.core.OperationType;
+import io.smallrye.graphql.client.dynamic.api.DynamicGraphQLClient;
+import io.smallrye.graphql.client.dynamic.api.DynamicGraphQLClientBuilder;
+import io.smallrye.graphql.client.typesafe.api.TypesafeGraphQLClientBuilder;
+
+/**
+ * This is a JAX-RS based wrapper that performs GraphQL client related tests.
+ * We can't perform these tests in the `@Test` methods directly, because the GraphQL client
+ * relies on CDI, and CDI is not available in native mode on the `@Test` side.
+ * Therefore the test only calls this REST endpoint which then performs all the client related work.
+ */
+@Path("/")
+public class GraphQLClientTester {
+
+    @GET
+    @Path("/typesafe/{url}")
+    public void typesafe(@PathParam("url") String url) {
+        LuckyNumbersClientApi client = TypesafeGraphQLClientBuilder.newBuilder()
+                .endpoint(url + "/graphql")
+                .build(LuckyNumbersClientApi.class);
+        client.setLuckyNumber(21);
+        Integer returned = client.luckyNumber();
+        if (!returned.equals(21)) {
+            throw new RuntimeException("Returned number was: " + returned);
+        }
+    }
+
+    @GET
+    @Path("/typesafe-header/{url}")
+    public void typesafeHeaders(@PathParam("url") String url) {
+        LuckyNumbersClientApiWithConfigKey client = TypesafeGraphQLClientBuilder.newBuilder()
+                .configKey("lucky")
+                .endpoint(url + "/graphql")
+                .build(LuckyNumbersClientApiWithConfigKey.class);
+        String header = client.returnHeader("myheader");
+        if (!"myvalue".equals(header)) {
+            throw new RuntimeException("Header 'myheader' was: " + header);
+        }
+    }
+
+    @GET
+    @Path("/typesafe-non-null/{url}")
+    public void typesafeNonNull(@PathParam("url") String url) {
+        LuckyNumbersClientApi client = TypesafeGraphQLClientBuilder.newBuilder()
+                .endpoint(url + "/graphql")
+                .build(LuckyNumbersClientApi.class);
+        List<Integer> someNumbers = List.of(12, 33);
+        List<Integer> returned = client.echoList(someNumbers);
+        if (!returned.equals(someNumbers)) {
+            throw new RuntimeException("Returned numbers were: " + returned);
+        }
+    }
+
+    @GET
+    @Path("/dynamic/{url}")
+    public void dynamic(@PathParam("url") String url) throws Exception {
+        try (DynamicGraphQLClient client = DynamicGraphQLClientBuilder.newBuilder().url(url + "/graphql").build()) {
+            Document setLuckyNumberMutation = document(
+                    operation(OperationType.MUTATION,
+                            field("set",
+                                    args(arg("newLuckyNumber", 15)))));
+            Document getLuckyNumberQuery = document(
+                    operation("number", field("get")));
+
+            // set the lucky number to 15
+            Response response = client.executeSync(setLuckyNumberMutation);
+            int returnedNumber = response.getData().getInt("set");
+            if (returnedNumber != 15) {
+                throw new RuntimeException("Unexpected response: " + response);
+            }
+
+            // get the lucky number and assert that it's 15
+            response = client.executeSync(getLuckyNumberQuery);
+            returnedNumber = response.getData().getInt("get");
+            if (returnedNumber != 15) {
+                throw new RuntimeException("Unexpected response: " + response);
+            }
+        }
+    }
+
+    @GET
+    @Path("/dynamic-subscription/{url}")
+    public void dynamicSubscription(@PathParam("url") String url) throws Exception {
+        try (DynamicGraphQLClient client = DynamicGraphQLClientBuilder.newBuilder().url(url.toString() + "/graphql").build()) {
+            Document op = document(
+                    operation(OperationType.SUBSCRIPTION,
+                            field("primeNumbers")));
+            List<Integer> expectedNumbers = List.of(2, 3, 5, 7, 11, 13);
+            List<Response> responses = client.subscription(op)
+                    .subscribe()
+                    .asStream()
+                    .collect(Collectors.toList());
+            for (int i = 0; i < expectedNumbers.size(); i++) {
+                if (expectedNumbers.get(i) != responses.get(i).getData().getInt("primeNumbers")
+                        || responses.get(i).hasError()) {
+                    throw new RuntimeException("Unexpected response: " + responses.get(i));
+                }
+            }
+        }
+    }
+
+}

--- a/integration-tests/smallrye-graphql-client/src/main/java/io/quarkus/io/smallrye/graphql/client/LuckyNumbersClientApiWithConfigKey.java
+++ b/integration-tests/smallrye-graphql-client/src/main/java/io/quarkus/io/smallrye/graphql/client/LuckyNumbersClientApiWithConfigKey.java
@@ -1,0 +1,25 @@
+package io.quarkus.io.smallrye.graphql.client;
+
+import java.util.List;
+
+import org.eclipse.microprofile.graphql.Mutation;
+import org.eclipse.microprofile.graphql.NonNull;
+import org.eclipse.microprofile.graphql.Query;
+
+import io.smallrye.graphql.client.typesafe.api.GraphQLClientApi;
+
+@GraphQLClientApi(configKey = "lucky")
+public interface LuckyNumbersClientApiWithConfigKey {
+
+    @Query(value = "get")
+    Integer luckyNumber();
+
+    @Mutation(value = "set")
+    Integer setLuckyNumber(Integer newLuckyNumber);
+
+    @Query(value = "echoList")
+    List<Integer> echoList(@NonNull List<Integer> list);
+
+    @Query
+    String returnHeader(String key);
+}

--- a/integration-tests/smallrye-graphql-client/src/main/java/io/quarkus/io/smallrye/graphql/client/LuckyNumbersResource.java
+++ b/integration-tests/smallrye-graphql-client/src/main/java/io/quarkus/io/smallrye/graphql/client/LuckyNumbersResource.java
@@ -3,12 +3,14 @@ package io.quarkus.io.smallrye.graphql.client;
 import java.util.List;
 
 import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
 
 import org.eclipse.microprofile.graphql.GraphQLApi;
 import org.eclipse.microprofile.graphql.Mutation;
 import org.eclipse.microprofile.graphql.NonNull;
 import org.eclipse.microprofile.graphql.Query;
 
+import io.quarkus.vertx.http.runtime.CurrentVertxRequest;
 import io.smallrye.graphql.api.Subscription;
 import io.smallrye.mutiny.Multi;
 
@@ -17,6 +19,9 @@ import io.smallrye.mutiny.Multi;
 public class LuckyNumbersResource {
 
     private volatile Integer luckyNumber = 12;
+
+    @Inject
+    CurrentVertxRequest request;
 
     @Query(value = "get")
     public Integer luckyNumber() {
@@ -37,6 +42,11 @@ public class LuckyNumbersResource {
     @Query(value = "echoList")
     public List<Integer> echoList(@NonNull List<Integer> list) {
         return list;
+    }
+
+    @Query
+    public String returnHeader(String key) {
+        return request.getCurrent().request().getHeader(key);
     }
 
 }

--- a/integration-tests/smallrye-graphql-client/src/main/resources/application.properties
+++ b/integration-tests/smallrye-graphql-client/src/main/resources/application.properties
@@ -1,0 +1,1 @@
+quarkus.smallrye-graphql-client.lucky.header.myheader=myvalue

--- a/integration-tests/smallrye-graphql-client/src/test/java/io/quarkus/it/smallrye/graphql/client/DynamicClientTest.java
+++ b/integration-tests/smallrye-graphql-client/src/test/java/io/quarkus/it/smallrye/graphql/client/DynamicClientTest.java
@@ -1,27 +1,17 @@
 package io.quarkus.it.smallrye.graphql.client;
 
-import static io.smallrye.graphql.client.core.Argument.arg;
-import static io.smallrye.graphql.client.core.Argument.args;
-import static io.smallrye.graphql.client.core.Document.document;
-import static io.smallrye.graphql.client.core.Field.field;
-import static io.smallrye.graphql.client.core.Operation.operation;
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertFalse;
+import static io.restassured.RestAssured.when;
 
 import java.net.URL;
-import java.util.List;
-import java.util.stream.Collectors;
 
 import org.junit.jupiter.api.Test;
 
 import io.quarkus.test.common.http.TestHTTPResource;
 import io.quarkus.test.junit.QuarkusTest;
-import io.smallrye.graphql.client.Response;
-import io.smallrye.graphql.client.core.Document;
-import io.smallrye.graphql.client.core.OperationType;
-import io.smallrye.graphql.client.dynamic.api.DynamicGraphQLClient;
-import io.smallrye.graphql.client.dynamic.api.DynamicGraphQLClientBuilder;
 
+/**
+ * See `GraphQLClientTester` for the actual testing code that uses GraphQL clients.
+ */
 @QuarkusTest
 public class DynamicClientTest {
 
@@ -29,41 +19,21 @@ public class DynamicClientTest {
     URL url;
 
     @Test
-    public void testDynamicClient() throws Exception {
-        try (DynamicGraphQLClient client = DynamicGraphQLClientBuilder.newBuilder().url(url.toString() + "/graphql").build()) {
-            Document setLuckyNumberMutation = document(
-                    operation(OperationType.MUTATION,
-                            field("set",
-                                    args(arg("newLuckyNumber", 15)))));
-            Document getLuckyNumberQuery = document(
-                    operation("number", field("get")));
-
-            // set the lucky number to 15
-            Response response = client.executeSync(setLuckyNumberMutation);
-            assertEquals(15, response.getData().getInt("set"), response.toString());
-
-            // get the lucky number and assert that it's 15
-            response = client.executeSync(getLuckyNumberQuery);
-            assertEquals(15, response.getData().getInt("get"), response.toString());
-        }
+    public void testDynamicClient() {
+        when()
+                .get("/dynamic/" + url.toString())
+                .then()
+                .log().everything()
+                .statusCode(204);
     }
 
     @Test
     public void testDynamicClientSubscription() throws Exception {
-        try (DynamicGraphQLClient client = DynamicGraphQLClientBuilder.newBuilder().url(url.toString() + "/graphql").build()) {
-            Document op = document(
-                    operation(OperationType.SUBSCRIPTION,
-                            field("primeNumbers")));
-            List<Integer> expectedNumbers = List.of(2, 3, 5, 7, 11, 13);
-            List<Response> responses = client.subscription(op)
-                    .subscribe()
-                    .asStream()
-                    .collect(Collectors.toList());
-            for (int i = 0; i < expectedNumbers.size(); i++) {
-                assertEquals(expectedNumbers.get(i), responses.get(i).getData().getInt("primeNumbers"));
-                assertFalse(responses.get(i).hasError());
-            }
-        }
+        when()
+                .get("/dynamic-subscription/" + url.toString())
+                .then()
+                .log().everything()
+                .statusCode(204);
     }
 
 }

--- a/integration-tests/smallrye-graphql-client/src/test/java/io/quarkus/it/smallrye/graphql/client/TypesafeClientTest.java
+++ b/integration-tests/smallrye-graphql-client/src/test/java/io/quarkus/it/smallrye/graphql/client/TypesafeClientTest.java
@@ -1,17 +1,17 @@
 package io.quarkus.it.smallrye.graphql.client;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
+import static io.restassured.RestAssured.when;
 
 import java.net.URL;
-import java.util.List;
 
 import org.junit.jupiter.api.Test;
 
-import io.quarkus.io.smallrye.graphql.client.LuckyNumbersClientApi;
 import io.quarkus.test.common.http.TestHTTPResource;
 import io.quarkus.test.junit.QuarkusTest;
-import io.smallrye.graphql.client.typesafe.api.TypesafeGraphQLClientBuilder;
 
+/**
+ * See `GraphQLClientTester` for the actual testing code that uses GraphQL clients.
+ */
 @QuarkusTest
 public class TypesafeClientTest {
 
@@ -19,24 +19,36 @@ public class TypesafeClientTest {
     URL url;
 
     @Test
-    public void testTypesafeClient() throws Exception {
-        LuckyNumbersClientApi client = TypesafeGraphQLClientBuilder.newBuilder()
-                .endpoint(url.toString() + "/graphql")
-                .build(LuckyNumbersClientApi.class);
-        client.setLuckyNumber(21);
-        assertEquals(21, client.luckyNumber());
+    public void testTypesafeClient() {
+        when()
+                .get("/typesafe/" + url.toString())
+                .then()
+                .log().everything()
+                .statusCode(204);
     }
 
     /**
      * Test a method that contains a `@NonNull` parameter.
      */
     @Test
-    public void testNonNull() throws Exception {
-        LuckyNumbersClientApi client = TypesafeGraphQLClientBuilder.newBuilder()
-                .endpoint(url.toString() + "/graphql")
-                .build(LuckyNumbersClientApi.class);
-        List<Integer> someNumbers = List.of(12, 33);
-        List<Integer> returned = client.echoList(someNumbers);
-        assertEquals(someNumbers, returned);
+    public void testNonNull() {
+        when()
+                .get("/typesafe-non-null/" + url.toString())
+                .then()
+                .log().everything()
+                .statusCode(204);
     }
+
+    /**
+     * Test HTTP headers defined in application.properties.
+     */
+    @Test
+    public void testHeader() {
+        when()
+                .get("/typesafe-header/" + url.toString())
+                .then()
+                .log().everything()
+                .statusCode(204);
+    }
+
 }

